official/resnet/resnet.py

```diff
@@ -169,9 +176,144 @@ def conv2d_fixed_padding(inputs, filters, kernel_size, strides, data_format):
+  then ReLU, then convolution. A schematic of the two architectures can be found
+  in Figure 1 (left) of:
+
+    Identity Mappings in Deep Residual Networks
+      https://arxiv.org/pdf/1603.05027.pdf
+      by Kaiming He, Xiangyu Zhang, Shaoqing Ren, and Jian Sun, Jul 2016.
+
+  Args:
+    version: Integer representing which version of the ResNet network to use.
+      See README for details. Valid values: [1, 2]
+    bottleneck: A boolean setting the use of bottleneck or building blocks.
+    inputs: A tensor of size [batch, channels, height_in, width_in] or
+      [batch, height_in, width_in, channels] depending on data_format.
+    filters: The number of filters for the convolutions.
+    training: A Boolean for whether the model is in training or inference
+      mode. Needed for batch normalization.
+    projection_shortcut: The function to use for projection shortcuts
+      (typically a 1x1 convolution when downsampling the input).
+    strides: The block's stride. If greater than 1, this block will ultimately
+      downsample the input.
+    data_format: The input format ('channels_last' or 'channels_first').
+
+  Returns:
+    The output tensor of the block.
+  """
+  block_dims = _block_dimensions(bottleneck=bottleneck, strides=strides)
+  shortcut = inputs
+
+  if version == 1:
+    block_dims[-1]["final"] = True  # Shortcut is added in the last layer.
+
+    if projection_shortcut is not None:
+      shortcut = projection_shortcut(inputs)
+      shortcut = batch_norm(inputs=shortcut, training=training,
+                            data_format=data_format)
+    for block in block_dims:
+      inputs = conv2d_fixed_padding(
+          inputs=inputs,
+          filters=filters * block["filters_mult"],
+          kernel_size=block["kernel_size"],
+          strides=block["strides"],
+          data_format=data_format)
+
+      inputs = batch_norm(inputs, training, data_format)
+      if block.get("final"):
+        inputs += shortcut
+      inputs = tf.nn.relu(inputs)
+
+    return inputs
+
+  elif version == 2:
+    block_dims[0]["first"] = True
+
+    inputs = batch_norm(inputs, training, data_format)
+    inputs = tf.nn.relu(inputs)
+
+    # The projection shortcut should come after the first batch norm and ReLU
+    # since it performs a 1x1 convolution.
+    if projection_shortcut is not None:
+      shortcut = projection_shortcut(inputs)
+
+    for block in block_dims:
+      # The first batch normalization and ReLU is performed outside of the loop
+      # in order to share operations with the projector where applicable.
+      if not block.get("first"):
+        inputs = batch_norm(inputs, training, data_format)
+        inputs = tf.nn.relu(inputs)
+
+      inputs = conv2d_fixed_padding(
+          inputs=inputs,
+          filters=filters * block["filters_mult"],
+          kernel_size=block["kernel_size"],
+          strides=block["strides"],
+          data_format=data_format)
+
+    return inputs + shortcut
```

Comment:
@karmel @nealwu I'm putting this forward as a plausible middle ground. By not having much duplication, we can be very liberal about what goes in docstrings and hopefully get clarity that way. But it unfortunately isn't as clear as the 4 function status quo and I think that means that we have to live with the duplicate code method. If you have strong feelings I'm willing to be persuaded otherwise though.

Reply:
This is interesting, but I think it really underscores that the readability of the four separate is preferable in this case. And this is coming from someone who goes long distances to get rid of duplicated code under most circumstances.

Reply:
Alright, I guess that settles it.